### PR TITLE
Fix insertion of child resources

### PR DESCRIPTION
**Pull request by alan-ghelardi**
This pull request resolves https://github.com/tektoncd/results/issues/296.

From now on the Watcher reads the record name from the
`results.tekton.dev/record` annotation only if the object in question is a
top-level resource, i.e. isn't owned by other objects. If the object is a
TaskRun associated to a PipelineRun, the `results.tekton.dev/record` annotation
is propagated from the PipelineRun to the TaskRun what causes conflicts:
ultimately, both controllers end up competing to update the same record. This
causes many precondition failed errors slowing down the reconciliation and
causing incongruities: one or more TaskRuns aren't inserted in the database
because those records are overwritten by the PipelineRun's record.

In addition, I included debug log messages to facilitate the investigation of
the Watcher's behavior.

A follow-up is creating E2E tests to prevent future regressions in this
behavior. I'd like to treat it in a future pull request.

### CODE REVIEW

**Comment by dibyom** [comment]
> If the object is a TaskRun associated to a PipelineRun, the results.tekton.dev/record annotation
is propagated from the PipelineRun to the TaskRun what causes conflicts

Ahh...this is an interesting edge case arising from our choice of propagating annotations from parent objects to child objects

**Review by dibyom — APPROVED** [review]
Thanks @alan-ghelardi - this looks good to me.

File: pkg/watcher/results/results.go [hunk 1]
```diff
@@ -181,9 +192,16 @@ func resultName(o metav1.Object) string {
...
-	name, ok := o.GetAnnotations()[annotation.Record]
-	if ok {
-		return name
+	// Attempt to read the record name from annotations only if the object
+	// in question is a top-level record (i.e. it isn't owned by another
+	// object). Otherwise, the annotation containing the record name maybe
+	// was propagated by the owner what causes conflicts while upserting the
+	// object into the API. For further details, please see
+	// https://github.com/tektoncd/results/issues/296.
+	if isTopLevelRecord(o) {
```
Comment (adambkaplan) [inline pkg/watcher/results/results.go:201]: :+1: on the descriptive comment.

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Review by khrm — COMMENTED** [review]
/kind bug
We need this label for release as discussed in the last WG call.

**Pull request merged**